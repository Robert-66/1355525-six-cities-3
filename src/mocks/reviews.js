export default [
  {
    comment: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    date: `2019-05-08T14:13:56.569Z`,
    id: 1,
    rating: 3,
    user: {
      avatarUrl: `img/1.png`,
      name: `Max`,
    }
  },
  {
    comment: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    date: `2017-08-08T14:13:56.569Z`,
    id: 2,
    rating: 5,
    user: {
      avatarUrl: `img/2.png`,
      name: `Anna`,
    }
  },
];
