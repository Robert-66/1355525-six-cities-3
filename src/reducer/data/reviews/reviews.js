import {extend} from '../../../utils';

const initialState = {
  data: [],
  isLoadingFetchReview: false,
  isLoadingCreateReview: false,
  isErrorFetchReview: false,
  isErrorCreateReview: false,
};

const ActionTypes = {
  FETCH_REVIEWS_START: `FETCH_REVIEWS_START`,
  FETCH_REVIEWS_SUCCESS: `FETCH_REVIEWS_SUCCESS`,
  FETCH_REVIEWS_FAILURE: `FETCH_REVIEWS_FAILURE`,
  CREATE_REVIEW_START: `CREATE_REVIEW_START`,
  CREATE_REVIEW_SUCCESS: `CREATE_REVIEW_SUCCESS`,
  CREATE_REVIEW_FAILURE: `CREATE_REVIEW_FAILURE`,
  RESET_REVIEW_STATE: `RESET_REVIEW_STATE`,
};

const ActionCreators = {
  fetchReviewsStart: () => ({
    type: ActionTypes.FETCH_REVIEWS_START,
  }),
  fetchReviewsSuccess: (data) => ({
    type: ActionTypes.FETCH_REVIEWS_SUCCESS,
    payload: data,
  }),
  fetchReviewsFailure: () => ({
    type: ActionTypes.FETCH_REVIEWS_FAILURE,
  }),
  createReviewStart: () => ({
    type: ActionTypes.CREATE_REVIEW_START,
  }),
  createReviewSucces: (data) => ({
    type: ActionTypes.CREATE_REVIEW_SUCCESS,
    payload: data
  }),
  createReviewFailure: () => ({
    type: ActionTypes.CREATE_REVIEW_FAILURE,
  }),
  resetReviewState: () => ({
    type: ActionTypes.RESET_REVIEW_STATE,
  })
};

const Operation = {
  fetchReviews: (offerId) => (dispatch, getState, api) => {
    dispatch(ActionCreators.fetchReviewsStart());
    return api.get(`/comments/${offerId}`)
      .then((response) => {
        dispatch(ActionCreators.fetchReviewsSuccess(response.data));
      })
      .catch(() => {
        dispatch(ActionCreators.fetchReviewsFailure());
      });
  },
  createReview: (offerId, data) => (dispatch, getState, api) => {
    return api.post(`/comments/${offerId}`, data)
      .then((response) => {
        dispatch(ActionCreators.createReviewSucces(response.data));
      })
      .catch(() => {
        dispatch(ActionCreators.createReviewFailure());
      });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypes.FETCH_REVIEWS_START:
      return extend(state, {isLoadingFetchReview: true, isErrorFetchReview: false});
    case ActionTypes.FETCH_REVIEWS_SUCCESS:
      return extend(state, {data: action.payload, isLoadingFetchReview: false});
    case ActionTypes.FETCH_REVIEWS_FAILURE:
      return extend(state, {data: [], isLoadingFetchReview: false, isErrorFetchReview: true});
    case ActionTypes.CREATE_REVIEW_START:
      return extend(state, {isLoadingCreateReview: true, isErrorCreateReview: false});
    case ActionTypes.CREATE_REVIEW_SUCCESS:
      return extend(state, {data: action.payload, isLoadingCreateReview: false});
    case ActionTypes.CREATE_REVIEW_FAILURE:
      return extend(state, {isLoadingCreateReview: false, isErrorCreateReview: true});
    case ActionTypes.RESET_REVIEW_STATE:
      return extend(state, {isLoadingCreateReview: false, isErrorCreateReview: false});
  }

  return state;
};

export {ActionCreators, ActionTypes, Operation, reducer};
