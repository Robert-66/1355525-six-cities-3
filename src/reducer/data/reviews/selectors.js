import {createSelector} from 'reselect';
import {extend} from '../../../utils';

function getTransformedReviews(reviews) {
  return reviews
    .slice(-10)
    .sort((a, b) => Date.parse(b.date) - Date.parse(a.date));
}

export const getReviews = createSelector(
    (state) => state.data.reviews,
    (reviews) => extend({}, {
      data: getTransformedReviews(reviews.data),
      isLoadingFetchReview: reviews.isLoadingFetchReview,
      isErrorFetchReview: reviews.isErrorFetchReview,
    })
);

export const getReviewsForm = createSelector(
    (state) => state.data.reviews,
    (reviews) => extend({}, {
      data: getTransformedReviews(reviews.data),
      isLoadingCreateReview: reviews.isLoadingCreateReview,
      isErrorCreateReview: reviews.isErrorCreateReview,
    })
);
