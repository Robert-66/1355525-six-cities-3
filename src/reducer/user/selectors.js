export function getAuthorizationStatus(state) {
  return state.user.authorizationStatus;
}
