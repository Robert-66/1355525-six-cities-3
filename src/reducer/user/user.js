import {extend} from '../../utils';

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  email: ``,
};

const ActionType = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const ActionCreator = {
  auth: (authInfo) => {
    return {
      type: ActionType.AUTH,
      payload: authInfo,
    };
  },
  noAuth: () => {
    return {
      type: ActionType.NO_AUTH,
    };
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.AUTH:
      return extend(state, {authorizationStatus: AuthorizationStatus.AUTH, email: action.payload.email});
    case ActionType.NO_AUTH:
      return initialState;
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((res) => {
        dispatch(ActionCreator.auth(res.data));
      });
  },
  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.email,
      password: authData.password,
    })
      .then((res) => {
        dispatch(ActionCreator.auth(res.data));
      });
  },
};


export {
  ActionCreator,
  ActionType,
  AuthorizationStatus,
  Operation,
  reducer,
};
