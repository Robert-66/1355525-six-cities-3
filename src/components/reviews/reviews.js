import React from 'react';
import ReviewsList from '../reviews-list/reviews-list';
import ReviewsForm from '../reviews-form/reviews-form';
import {reviewType} from '../../types/reviews-types.js';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {AuthorizationStatus} from '../../reducer/user/user';
import {ActionCreators} from '../../reducer/data/reviews/reviews';
import {Operation as ReviewsOperation} from "../../reducer/data/reviews/reviews";
import {getReviews, getReviewsForm} from "../../reducer/data/reviews/selectors";

class Reviews extends React.PureComponent {
  constructor(props) {
    super(props);

    this.handleReviewSubmit = this.handleReviewSubmit.bind(this);
  }

  componentDidMount() {
    this.props.fetchReviews(this.props.offerId);
  }

  handleReviewSubmit(reviewData) {
    this.props.onSubmitReview(this.props.offerId, reviewData);
  }

  render() {
    const {
      className,
      reviews,
      reviewsForm,
      authorizationStatus,
      onResetReviewFormState,
    } = this.props;

    return (
      <section className={`reviews${className ? ` ` + className : ``}`}>
        {reviews.data.length > 0 && (
          <>
            <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviews.data.length}</span></h2>
            <ReviewsList reviews={reviews.data}/>
            {authorizationStatus === AuthorizationStatus.AUTH && (
              <ReviewsForm
                reviews={reviewsForm}
                onSubmit={this.handleReviewSubmit}
                onResetState={onResetReviewFormState}
              />
            )}
          </>
        )}
      </section>
    );
  }
}

Reviews.propTypes = {
  className: PropTypes.string,
  reviews: PropTypes.shape({
    data: PropTypes.arrayOf(reviewType).isRequired,
    isLoadingFetchReview: PropTypes.bool.isRequired,
    isErrorFetchReview: PropTypes.bool.isRequired,
  }).isRequired,
  reviewsForm: PropTypes.shape({
    data: PropTypes.arrayOf(reviewType).isRequired,
    isLoadingCreateReview: PropTypes.bool.isRequired,
    isErrorCreateReview: PropTypes.bool.isRequired,
  }).isRequired,
  authorizationStatus: PropTypes.string.isRequired,
  offerId: PropTypes.number.isRequired,
  onSubmitReview: PropTypes.func.isRequired,
  onResetReviewFormState: PropTypes.func.isRequired,
  fetchReviews: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    reviews: getReviews(state),
    reviewsForm: getReviewsForm(state),
    authorizationStatus: state.user.authorizationStatus,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    fetchReviews: (offerId) => dispatch(ReviewsOperation.fetchReviews(offerId)),
    onSubmitReview: (offerId, reviewData) => dispatch(ReviewsOperation.createReview(offerId, reviewData)),
    onResetReviewFormState: () => dispatch(ActionCreators.resetReviewState()),
  };
}

export {Reviews};
export default connect(mapStateToProps, mapDispatchToProps)(Reviews);
