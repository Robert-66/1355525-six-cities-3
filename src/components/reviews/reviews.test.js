import React from 'react';
import renderer from 'react-test-renderer';
import {AuthorizationStatus} from '../../reducer/user/user';
import {Reviews} from './reviews';

const mocks = {
  reviews: [
    {
      comment: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
      date: `2019-05-08T14:13:56.569Z`,
      id: 1,
      rating: 3,
      user: {
        avatarUrl: `img/1.png`,
        name: `Max`,
      }
    },
    {
      comment: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
      date: `2017-08-08T14:13:56.569Z`,
      id: 2,
      rating: 5,
      user: {
        avatarUrl: `img/2.png`,
        name: `Anna`,
      }
    },
  ],
  reviewsForm: {
    data: [],
    isLoading: false,
    isError: false,
  }
};

describe(`Reviews is rendered correctly`, () => {

  it(`when the user is authorized`, () => {
    const tree = renderer
      .create(
          <Reviews
            reviews={mocks.reviews}
            reviewsForm={mocks.reviewsForm}
            authorizationStatus={AuthorizationStatus.AUTH}
            offerId={1}
            onSubmitReview={() => {}}
            onResetReviewFormState={() => {}}
          />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`when the user is unauthorized`, () => {
    const tree = renderer
      .create(
          <Reviews
            reviews={mocks.reviews}
            reviewsForm={mocks.reviewsForm}
            authorizationStatus={AuthorizationStatus.NO_AUTH}
            offerId={1}
            onSubmitReview={() => {}}
            onResetReviewFormState={() => {}}
          />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
