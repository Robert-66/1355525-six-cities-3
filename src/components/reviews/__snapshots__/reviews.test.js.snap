// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews is rendered correctly when the user is authorized 1`] = `
<section
  className="reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews · 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="img/1.png"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "60%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
        </p>
        <time
          className="reviews__time"
          dateTime="2019-05-08"
        >
          May 2019
        </time>
      </div>
    </li>
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="img/2.png"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "100%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
        </p>
        <time
          className="reviews__time"
          dateTime="2017-08-08"
        >
          August 2017
        </time>
      </div>
    </li>
  </ul>
  <form
    action="#"
    className="reviews__form form"
    onSubmit={[Function]}
  >
    <label
      className="reviews__label form__label"
      htmlFor="review"
    >
      Your review
    </label>
    <div
      className="reviews__rating-form form__rating"
    >
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        disabled={false}
        id="5-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={5}
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="5-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        disabled={false}
        id="4-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={4}
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="4-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        disabled={false}
        id="3-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={3}
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="3-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        disabled={false}
        id="2-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={2}
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="2-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        disabled={false}
        id="1-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={1}
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="1-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
    </div>
    <textarea
      className="reviews__textarea form__textarea"
      disabled={false}
      id="review"
      name="review"
      onChange={[Function]}
      placeholder="Tell how was your stay, what you like and what can be improved"
      value=""
    />
    <div
      className="reviews__button-wrapper"
    >
      <p
        className="reviews__help"
      >
        To submit review please make sure to set 
        <span
          className="reviews__star"
        >
          rating
        </span>
         and describe your stay with at least 
        <b
          className="reviews__text-amount"
        >
          50 characters
        </b>
        .
      </p>
      <button
        className="reviews__submit form__submit button"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</section>
`;

exports[`Reviews is rendered correctly when the user is unauthorized 1`] = `
<section
  className="reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews · 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="img/1.png"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "60%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
        </p>
        <time
          className="reviews__time"
          dateTime="2019-05-08"
        >
          May 2019
        </time>
      </div>
    </li>
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="img/2.png"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "100%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
        </p>
        <time
          className="reviews__time"
          dateTime="2017-08-08"
        >
          August 2017
        </time>
      </div>
    </li>
  </ul>
</section>
`;
